pkg/resolve/selector.go

```diff
@@ -0,0 +1,116 @@
+package resolve
+
+import (
+	"bytes"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	yaml2json "k8s.io/apimachinery/pkg/util/yaml"
+	"regexp"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+
+// FilterBySelector filters out any resources
+// from the raw manifest bytes whose labels
+// don't match the provided selector
+func FilterBySelector(input []byte, selectorString string) ([]byte, error) {
+	selector, err := labels.Parse(selectorString)
+	if err != nil {
+		return nil, err
+	}
+
+	var outputObjectsYaml [][]byte
+
+	// helper function to append selected objects to our result
+	appendObj := func(obj runtime.Object) error {
+		rawJson, err := runtime.Encode(unstructured.UnstructuredJSONScheme, obj)
+		if err != nil {
+			return err
+		}
+		rawYaml, err := yaml.JSONToYAML(rawJson)
+		if err != nil {
+			return err
+		}
+		outputObjectsYaml = append(outputObjectsYaml, rawYaml)
+		return nil
+	}
+
+	// parse runtime.Objects from the input yaml
+	objects, err := parseUnstructured(input)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, object := range objects {
+		// objects parsed by UnstructuredJSONScheme can only be of
+		// type *unstructured.Unstructured or *unstructured.UnstructuredList
+		switch unstructuredObj := object.(type) {
+		case *unstructured.Unstructured:
+			// append the object if it matches the provided labels
+			if selector.Matches(labels.Set(unstructuredObj.GetLabels())) {
+				if err := appendObj(unstructuredObj); err != nil {
+					return nil, err
+				}
+			}
+		case *unstructured.UnstructuredList:
+			// filter the list items based on label
+			var filteredItems []unstructured.Unstructured
+			for _, obj := range unstructuredObj.Items {
+				if selector.Matches(labels.Set(obj.GetLabels())) {
+					filteredItems = append(filteredItems, obj)
+				}
+			}
+			// only append the list if it still contains items after being filtered
+			if len(filteredItems) > 0 {
+				unstructuredObj.Items = filteredItems
+				if err := appendObj(unstructuredObj); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	// re-join the objects into a single manifest
+	return bytes.Join(outputObjectsYaml, []byte("---\n")), nil
+}
+
+var yamlSeparatorRegex = regexp.MustCompile( "\n---")
+
+func parseUnstructured(rawYaml []byte) ([]runtime.Object, error) {
+	objectYamls := yamlSeparatorRegex.Split(string(rawYaml), -1)
```

Comment:
You could use [this](https://github.com/kubernetes/apimachinery/blob/ed135c5b96450fd24e5e981c708114fbbd950697/pkg/util/yaml/decoder.go#L93) for splitting the document -- I'm not sure if it will look any cleaner, though.

Reply:
the problem is that this won't let me use `runtime.Decode(unstructured.UnstructuredJSONScheme) `